File: src/csharp/pulse/cdm/io/PBEvents.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

using System.Collections.Generic;
using Google.Protobuf;

namespace Pulse.CDM
{
  public class PBEvents
  {
    public static void Load(pulse.cdm.bind.ActiveEventListData src, List<SEActiveEvent> dst)
    {
      Serialize(src, dst);
    }
    public static void Serialize(pulse.cdm.bind.ActiveEventListData src, List<SEActiveEvent> dst)
    {
      SEScalarTime duration = new SEScalarTime();
      foreach (var active in src.ActiveEvent)
      {
        PBProperty.Load(active.Duration, duration);
        dst.Add(new SEActiveEvent((eEvent)active.Event, duration));
      }
    }
    public static bool SerializeFromString(string src, List<SEActiveEvent> dst)
    {
      try
      {
        pulse.cdm.bind.ActiveEventListData data = JsonParser.Default.Parse<pulse.cdm.bind.ActiveEventListData>(src);
        PBEvents.Load(data, dst);
      }
      catch (Google.Protobuf.InvalidJsonException)
      {
        return false;
      }
      return true;
    }

    public static pulse.cdm.bind.ActiveEventListData Unload(List<SEActiveEvent> src)
    {
      pulse.cdm.bind.ActiveEventListData dst = new pulse.cdm.bind.ActiveEventListData();
      Serialize(src, dst);
      return dst;
    }
    public static void Serialize(List<SEActiveEvent> src, pulse.cdm.bind.ActiveEventListData dst)
    {
      foreach (var active in src)
      {
        pulse.cdm.bind.ActiveEventData ae = new pulse.cdm.bind.ActiveEventData();
        ae.Event = (pulse.cdm.bind.eEvent)active.Event;
        ae.Duration = PBProperty.Unload(active.Duration);
        dst.ActiveEvent.Add(ae);
      }
    }
    public static string SerializeToString(List<SEActiveEvent> src)
    {
      var pb = PBEvents.Unload(src);
      return pb.ToString();
    }

    public static void Load(pulse.cdm.bind.EventChangeListData src, List<SEEventChange> dst)
    {
      Serialize(src, dst);
    }
    public static void Serialize(pulse.cdm.bind.EventChangeListData src, List<SEEventChange> dst)
    {
      SEScalarTime simTime = new SEScalarTime();
      foreach (var change in src.Change)
      {
        PBProperty.Load(change.SimTime, simTime);
        dst.Add(new SEEventChange((eEvent)change.Event, change.Active, simTime));
      }
    }
    public static bool SerializeFromString(string src, List<SEEventChange> dst)
    {
      try
      {
        pulse.cdm.bind.EventChangeListData data = JsonParser.Default.Parse<pulse.cdm.bind.EventChangeListData>(src);
        PBEvents.Load(data, dst);
      }
      catch (Google.Protobuf.InvalidJsonException)
      {
        return false;
      }
      return true;
    }

    public static pulse.cdm.bind.EventChangeListData Unload(List<SEEventChange> src)
    {
      pulse.cdm.bind.EventChangeListData dst = new pulse.cdm.bind.EventChangeListData();
      Serialize(src, dst);
      return dst;
    }
    public static void Serialize(List<SEEventChange> src, pulse.cdm.bind.EventChangeListData dst)
    {
      foreach (var change in src)
      {
        pulse.cdm.bind.EventChangeData c = new pulse.cdm.bind.EventChangeData();
        c.Event = (pulse.cdm.bind.eEvent)change.Event;
        c.Active = change.Active;
        if (change.SimTime.IsValid())
          c.SimTime = PBProperty.Unload(change.SimTime);
        dst.Change.Add(c);
      }
    }
    public static string SerializeToString(List<SEEventChange> src)
    {
      var pb = PBEvents.Unload(src);
      return pb.ToString();
    }
  }
}

File: src/cpp/cdm/io/protobuf/PBEvents.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
PUSH_PROTO_WARNINGS()
#include "pulse/cdm/bind/Events.pb.h"
POP_PROTO_WARNINGS()
#include "io/protobuf/PBEvents.h"
#include "io/protobuf/PBProperties.h"
#include "io/protobuf/PBUtils.h"
#include "engine/SEEventManager.h"
#include "utils/FileUtils.h"

const std::string& eEvent_Name(eEvent m)
{
  return CDM_BIND::eEvent_Name((CDM_BIND::eEvent)m);
}

bool PBEvents::SerializeToString(std::vector<const SEEventChange*>& changes, std::string& output, SerializationFormat m, Logger* logger)
{
  if (changes.empty())
    return false;

  CDM_BIND::EventChangeListData data;
  for (const SEEventChange* c : changes)
  {
    CDM_BIND::EventChangeData* change = data.add_change();
    change->set_event((CDM_BIND::eEvent)c->GetEvent());
    change->set_active(c->GetActive());
    if (c->GetSimTime().IsValid())
      change->set_allocated_simtime(PBProperty::Unload(c->GetSimTime()));
  }
  if (!PBUtils::SerializeToString(data, output, m, logger))
    return false;
  return true;
}

bool PBEvents::SerializeFromString(const std::string& src, std::vector<const SEEventChange*>& changes, SerializationFormat m, Logger* logger)
{
  SEScalarTime time;
  CDM_BIND::EventChangeListData data;
  if (!PBUtils::SerializeFromString(src, data, m, logger))
    return false;
  for (int i = 0; i < data.change_size(); i++)
  {
    auto& e = data.change()[i];
    if (e.has_simtime())
      PBProperty::Load(e.simtime(), time);
    else
      time.Invalidate();
    SEEventChange* c = new SEEventChange((eEvent)e.event(), e.active(), &time);
    changes.push_back(c);
  }

  return true;
}

bool PBEvents::SerializeToString(std::vector<const SEActiveEvent*>& active_events, std::string& output, SerializationFormat m, Logger* logger)
{
  if (active_events.empty())
    return false;

  CDM_BIND::ActiveEventListData data;
  for (const SEActiveEvent* a : active_events)
  {
    CDM_BIND::ActiveEventData* aeData = data.add_activeevent();
    aeData->set_event((CDM_BIND::eEvent)a->GetEvent());
    if (a->GetDuration().IsValid())
      aeData->set_allocated_duration(PBProperty::Unload(a->GetDuration()));
  }
  if (!PBUtils::SerializeToString(data, output, m, logger))
    return false;
  return true;
}

bool PBEvents::SerializeFromString(const std::string& src, std::vector<const SEActiveEvent*>& active_events, SerializationFormat m, Logger* logger)
{
  SEScalarTime time;
  CDM_BIND::ActiveEventListData data;
  if (!PBUtils::SerializeFromString(src, data, m, logger))
    return false;
  for (int i = 0; i < data.activeevent_size(); i++)
  {
    auto& ae = data.activeevent()[i];
    if (ae.has_duration())
      PBProperty::Load(ae.duration(), time);
    else
      time.Invalidate();
    SEActiveEvent* c = new SEActiveEvent((eEvent)ae.event(), time);
    active_events.push_back(c);
  }

  return true;
}